Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,373 @@
+                if decoded.contains(character) {
+                    throw MongoError.InvalidArgumentError(
+                        message: "defaultAuthDB contains invalid character: \(character)"
+                    )
+                }
+            }
+            self.defaultAuthDB = decoded
+            // If no other authentication options were provided, we should use the defaultAuthDB as the credential
+            // source. This will be overwritten later if an authSource is provided.
+            self.credential?.source = decoded
+        }
+
+        guard authDatabaseAndOptions.count == 2 else {
+            // No options were specified
+            return
+        }
+
+        let options = try Options(authDatabaseAndOptions[1])
+
+        // Parse authentication options into a MongoCredential
+        if let mechanism = options.authMechanism {
+            var credential = self.credential ?? MongoCredential()
+            credential.source = try options.authSource ?? mechanism.getDefaultSource(defaultAuthDB: self.defaultAuthDB)
+            credential.mechanismProperties = options.authMechanismProperties
+            try mechanism.validateAndUpdateCredential(credential: &credential)
+            credential.mechanism = mechanism
+            self.credential = credential
+        } else if options.authMechanismProperties != nil {
+            throw MongoError.InvalidArgumentError(
+                message: "Connection string specified authMechanismProperties but no authMechanism was specified"
+            )
+        } else if let authSource = options.authSource {
+            // If no mechanism was provided but authentication was requested and an authSource was provided, populate
+            // the source field with the authSource. Otherwise, source will fall back to the defaultAuthDB if provided,
+            // or "admin" if not.
+            if self.credential != nil {
+                self.credential?.source = authSource
+            } else {
+                // The authentication mechanism defaults to SCRAM if an authMechanism is not provided, and SCRAM
+                // requires a username
+                throw MongoError.InvalidArgumentError(
+                    message: "No username provided for authentication in the connection string but an authSource was"
+                        + " provided. To use an authentication mechanism that does not require a username, specify an"
+                        + " authMechanism in the connection string."
+                )
+            }
+        }
+    }
+
+    /// `Codable` conformance
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(self.description)
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let stringValue = try container.decode(String.self)
+        try self.init(throwsIfInvalid: stringValue)
```

Comment:
We'll want to convert this error to a `DecodingError` to preserve the decoding metadata (e.g. the key path) for debugging. You can see an example of that here: https://github.com/mongodb/mongo-swift-driver/blob/main/Sources/MongoSwift/MongoServerAPI.swift#L46.